doc/neps/nep-0053-c-abi-evolution.md

```diff
@@ -0,0 +1,313 @@
+
+The implementation of this NEP consists would consist of two steps:
+1. As part of a general improvement, starting with NumPy 1.25 building with
+   NumPy will by default export an older API version to allow backwards
+   compatible builds with the newest available NumPy version.
+   (New API is not available unless opted-in.)
+2. The NumPy 2.0 will:
+   * require recompilation of downstream packages against NumPy 2.0 to be
+     compatible with NumPy 2.0.
+   * need a ``numpy2_compat`` as a dependency when running on NumPy 1.x.
+   * require some downstream code changes to adapt to changed API.
+
+
+Motivation and Scope
+====================
+
+The NumPy API conists of more than 300 functions and numerous macros.
+Many of these are outdated: some were only ever used within NumPy,
+exist only for compatibility with NumPy's predecessors, or have no or only
+a single known downstream user (i.e. SciPy).
+
+Further, many structs used by NumPy have always been public making it
+impossible to change them outside of a major release.
+Some changes have been planned for years and were the reason for
+``NPY_NO_DEPRECATED_API`` and further deprecations as explained in
+`c_api_deprecations`_.
+
+While we probably have little reason to change the layout of the array struct
+(``PyArrayObject_fields``) for example the development and improvement of
+dtypes would be made easier by changing the `PyArray_Descr` struct.
+
+This NEP proposes a few concrete changes to our C-API mainly as examples.
+However, more changes will be handled on a case-by-case basis, and we do not
+aim to provide a full list of changes in this NEP.
+
+Adding state is out of scope
+----------------------------
+New developements such as CPython's support for subinterpreters and the
+HPy API may require the NumPy C-API to evolve in a way that may require
+(or at least prefer) state to be passed in.
+
+As of now, we do not aim to include changes for this here.  We cannot expect
+users to do large code updates to pass in for example an ``HPy`` context
+to many NumPy functions.
+
+While we could introduce a second API for this purpose in NumPy 2.0,
+we expect that this is unnecessary and that the provisions introduced here:
+
+* the ability to compile with the newest NumPy version but be compatible with
+  older versions,
+* and the possibility of updating a ``numpy2_compat`` package.
+
+should allow to add such an API also in a minor release.
+
+
+Usage and Impact
+================
+
+Backwards compatible builds
+---------------------------
+
+Backwards compatible builds will be described in more details in the
+documentation.
+Briefly, we will allow users to use a definition like::
+
+    #define NPY_TARGET_VERSION NPY_1_22_API_VERSION
+
+to select the version they wish to compile for (lowest version to be
+compatible with).
+This will default to an old version that should cover most or all use-cases.
+The version will be chosen to always be compatible with :ref:`NEP 29 <NEP29>`.
```

Comment:
As a follow-up (to the discussion above).  I am currently planning to change this away from NEP 29 again to a very conservative default based on "oldest NumPy compatible with the oldest Python version the current NumPy supports" (that parses terribly).

That would make things simpler (I think), because you don't have to worry that a new minor release of SciPy will accidentally drop support for an old version because you compile against a newer NumPy version (which should be fine).